runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java

```diff
@@ -0,0 +1,435 @@
+          baseContext
+              .windowingInternals()
+              .outputWindowedValue(output, timestamp, element.getWindows(), element.getPane());
+          noteOutput();
+        }
+
+        private void noteOutput() {
+          if (++numOutputs >= MAX_OUTPUTS_PER_BUNDLE) {
+            // Request a checkpoint. The fn *may* produce more output, but hopefully not too much.
+            residualRestrictionHolder[0] = tracker.checkpoint();
+          }
+        }
+
+        public <T> T sideInput(PCollectionView<T> view) {
+          return baseContext.sideInput(view);
+        }
+
+        public PipelineOptions getPipelineOptions() {
+          return baseContext.getPipelineOptions();
+        }
+
+        public <T> void sideOutput(TupleTag<T> tag, T output) {
+          // TODO: I'm not sure how to implement this correctly: there's no
+          // "internals.sideOutputWindowedValue".
+          throw new UnsupportedOperationException(
+              "Side outputs not yet supported by splittable DoFn");
+        }
+
+        public <T> void sideOutputWithTimestamp(TupleTag<T> tag, T output, Instant timestamp) {
+          // TODO: I'm not sure how to implement this correctly: there's no
+          // "internals.sideOutputWindowedValue".
+          throw new UnsupportedOperationException(
+              "Side outputs not yet supported by splittable DoFn");
+        }
+      };
+    }
+
+    /** Creates an {@link DoFn.ExtraContextFactory} that provides just the given tracker. */
+    private DoFn.ExtraContextFactory<InputT, OutputT> wrapTracker(final TrackerT tracker) {
+      return new DoFn.ExtraContextFactory<InputT, OutputT>() {
+        @Override
+        public BoundedWindow window() {
+          // DoFnSignatures should have verified that this DoFn doesn't access extra context.
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
+        }
+
+        @Override
+        public DoFn.InputProvider<InputT> inputProvider() {
+          // DoFnSignatures should have verified that this DoFn doesn't access extra context.
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
+        }
+
+        @Override
+        public DoFn.OutputReceiver<OutputT> outputReceiver() {
+          // DoFnSignatures should have verified that this DoFn doesn't access extra context.
+          throw new IllegalStateException("Unexpected extra context access on a splittable DoFn");
+        }
+
+        @Override
+        public TrackerT restrictionTracker() {
+          return tracker;
+        }
+      };
+    }
+  }
+
+  /** Splits the restriction using the given {@link DoFn.SplitRestriction} method. */
+  private static class SplitRestrictionFn<InputT, RestrictionT>
+      extends DoFn<
+          ElementRestriction<InputT, RestrictionT>, ElementRestriction<InputT, RestrictionT>> {
+    private final DoFn<InputT, ?> fn;
```

Comment:
maybe call this splittableFn or something to make it clear which `fn` we're referring to?